syntax = "proto3";
package industry.decisions.proto;

import "games/market/proto/goods.proto";

message VariantInfo {
  // Cost per unit of consumable inputs.
  int64 unit_cost_u = 1;
  // Maximum scale.
  int64 possible_scale_u = 2;
  // Resource that makes the limit on this variant.
  string bottleneck = 3;
  // Cost at scale (not per unit!) of capital, including install costs.
  int64 cap_cost_u = 4;
}

message StepInfo {
  repeated VariantInfo variant = 1;
  int32 best_variant = 2;
}

message ProductionInfo {
  string name = 1;
  int64 max_scale_u = 2;
  int64 total_unit_cost_u = 3;
  repeated StepInfo step_info = 4;
  string reject_reason = 5;
  market.proto.Container expected_output = 6;
}

// Holds information about which production chain was
// selected and why.
message ProductionDecision {
  // The best chain, if there is one.
  ProductionInfo selected = 1;
  // Chains that are possible, but not as good as the selected one.
  repeated ProductionInfo rejected = 2;
}
